packages/langium/src/parser/token-builder.ts

```diff
@@ -4,58 +4,37 @@
 import { stream } from '../utils/stream';
 
 export interface TokenBuilder {
-    buildTokens(grammar: Grammar, options?: { caseInsensitive?: boolean }): TokenType[];
+    buildTokens(grammar: Grammar, options?: { caseInsensitive?: boolean }): TokenVocabulary;
 }
 
 export class DefaultTokenBuilder implements TokenBuilder {
 
-    // We need suffixes for terminals and keywords which have the same name
-    protected readonly KEYWORD_SUFFIX = '_KEYWORD';
-    protected readonly TERMINAL_SUFFIX = '_TERMINAL';
+    buildTokens(grammar: Grammar, options?: { caseInsensitive?: boolean }): TokenVocabulary {
+        const terminalTokens: TokenType[] = this.buildTerminalTokens(grammar);
+        const tokens: TokenType[] = this.buildKeywordTokens(grammar, terminalTokens, options);
 
-    buildTokens(grammar: Grammar, options?: { caseInsensitive?: boolean }): TokenType[] {
-        const tokenMap = new Map<string, TokenType>();
-        const terminalsTokens: TokenType[] = [];
-        const terminals = Array.from(stream(grammar.rules).filter(isTerminalRule)).filter(e => !e.fragment);
-        for (const terminal of terminals) {
-            const token = this.buildTerminalToken(terminal);
-            terminalsTokens.push(token);
-            tokenMap.set(terminal.name + this.TERMINAL_SUFFIX, token);
-        }
-
-        const tokens: TokenType[] = [];
-        // We filter by parser rules, since keywords in terminal rules get transformed into regex and are not actual tokens
-        const parserRuleKeywords = grammar.rules.filter(isParserRule).flatMap(rule => streamAllContents(rule).filter(isKeyword).toArray());
-        const keywords = stream(parserRuleKeywords).distinct(e => e.value).toArray()
-            // Sort keywords by descending length
-            .sort((a, b) => b.value.length - a.value.length);
-
-        for (const keyword of keywords) {
-            const keywordToken = this.buildKeywordToken(keyword, keywords, terminals, tokenMap, !!options?.caseInsensitive);
-            tokens.push(keywordToken);
-            tokenMap.set(keyword.value + this.KEYWORD_SUFFIX, keywordToken);
-        }
-
-        for (const terminalToken of terminalsTokens) {
+        terminalTokens.forEach(terminalToken => {
             const pattern = terminalToken.PATTERN;
-            if (typeof pattern === 'object' && pattern && 'test' in pattern && pattern.test(' ')) {
-                tokens.unshift(terminalToken);
-            } else {
+            (typeof pattern === 'object' && pattern && 'test' in pattern && pattern.test(' ')) ?
+                tokens.unshift(terminalToken) :
                 tokens.push(terminalToken);
-            }
-        }
-
+        });
         return tokens;
     }
 
+    protected buildTerminalTokens(grammar: Grammar): TokenType[] {
+        return Array.from(stream(grammar.rules).filter(isTerminalRule)).filter(e => !e.fragment)
+            .map(terminal => this.buildTerminalToken(terminal));
```

Comment:
Yes, it's a nice idea! However, `buildTerminalTokens` and `buildKeywordTokens` can be used in custom implementations of `TokenBuilder`, and it's nice if they both return or streams, or arrays. An array is more natural for `buildKeywordTokens` (moreover, it's a base for the returning result that have to be an array), so, I'd propose to use an array for `buildTerminalTokens` too.

Reply:
Ok but even then you can defer array creation to the end to avoid creating intermediate arrays:
```typescript
return stream(grammar.rules).filter(isTerminalRule).filter(e => !e.fragment)
    .map(terminal => this.buildTerminalToken(terminal))
    .toArray();
```